src/kernel/sporadic.c

```diff
@@ -210,135 +214,84 @@ void refill_update(sched_context_t *sc, ticks_t new_period, ticks_t new_budget,
 
 static inline void schedule_used(sched_context_t *sc, refill_t new)
 {
-    /* schedule the used amount */
-    if (new.rAmount < MIN_BUDGET && !refill_single(sc)) {
-        /* used amount is to small - merge with last and delay */
-        refill_tail(sc)->rAmount += new.rAmount;
-        refill_tail(sc)->rTime = MAX(new.rTime, refill_tail(sc)->rTime);
-    } else if (new.rTime <= refill_tail(sc)->rTime) {
+    if (unlikely(refill_tail(sc)->rTime + refill_tail(sc)->rAmount >= new.rTime)) {
```

Comment:
I find this situation difficult to picture. How can a new refill overlap with an existing one? Wouldn't that mean that whatever caused that earlier replenishment to be scheduled should still be running?

Reply:
The only case where overlap should occur is after a `refill_update`, where the second of the two refills _starts_ one period after the start of the head refill. In this case, any time charged to the head refill will be placed one period later overlapping the tail refill.

This case also includes _adjacent_ refills, such as when we only partially charge a refill at one point and then later charge a fragment from what would have been the same refill. If that refill was not delayed (such as in the case of preemption) then the second time we charge that refill, the replenishment will start immediately when the charge from the previous replenishment ended. This ensures that the _only_ case where more than two refills is actually used is when a thread actively blocks without exhausting all available time, as this causes the refill to be delayed before it is next charged so the next charge cannot be merged with the replenishment before.

Reply:
Ok, I think I get it now. On first sight it seems a bit convoluted, but I guess it makes sense given everything else. Thanks for the explanation. (Sounding like a broken record, this is also something I would suggest to note in a comment for posterity.)

Reply:
Happy to add more comments for clarity. I'm not sure there is a significant reason for the general lack of comments in the C kernel aside from the means by which it was originally translated from the Haskell implementation (which is substantially better commented).